### PR TITLE
call meshChanged() earlier

### LINKED ISSUE #15934: EqualValueBoundaryConstraint doesn't work with adaptivity
## Bug Description

The EqualValueBoundaryConstraint is only enforced on-processor when there is adaptivity, or more generally anything that triggers meshChanged() will break it.

## Steps to Reproduce

Solve a problem with EqualValueBoundaryConstraint and adaptivity.

## Impact

This bug prevents EqualValueBoundaryConstraint to be used in parallel with adaptivity, XFEM, etc..

### PR DESCRIPTION
**Pull request by hugary1995**
This allows the ghosting inside EqualValueBoundaryConstraint::meshChanged() to take effect.

close #15934

### CODE REVIEW

**Comment by hugary1995** [comment]
Without this patch, wrong result in parallel (the constraint is only enforced on-processor):

With this patch:


For details please refer to the test I added.

@lindsayad this is the change I was mentioning earlier on slack, could you please review this? I understand we are moving towards RM-based constraint system, but I need to fix this bug now for a project (given that this change doesn't break existing tests).

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -6091,6 +6091,9 @@ FEProblemBase::meshChangedHelper(bool intermediate_change)
   // repartitioning done in EquationSystems::reinit().
   _mesh.meshChanged();
 
+  for (const auto & mci : _notify_when_mesh_changes)
+    mci->meshChanged();
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:6095]: It would be great if you could put some code comment here on why this has to be where it is

**Review by lindsayad — CHANGES_REQUESTED** [review]
Just the code comment request. And might as well squash the fixup commit into the original commit (in the future you can make use of `git commit --fixup` and `git rebase --autosquash` if you weren't aware of those commands)

**Comment by hugary1995** [comment]
Thanks Alex, these commands are useful. I added a comment explaining why I had to do this.
I also restricted the test to replicated mesh only -- I am not sure if that is cheating but I see some precedents in some other constraint tests.

**Comment by lindsayad** [comment]
> I also restricted the test to replicated mesh only -- I am not sure if that is cheating but I see some precedents in some other constraint tests.

Was your test failing in distributed mode?

**Comment by hugary1995** [comment]
Yes it was failing some distributed mesh tests.

**Comment by lindsayad** [comment]
> Yes it was failing some distributed mesh tests.

Ok we won't make you figure out the root cause of that. You didn't alter any existing test specs, so you didn't break any existing distributed mesh tests with this change.

**Review by lindsayad — APPROVED** [review]

**Pull request merged**